File: wolf/src/stream/rist.rs
#![allow(improper_ctypes)]

use crate::stream::ffi::rist::{
    w_receiver_data_callback, w_rist_auth_connect_callback, w_rist_auth_disconnect_callback,
    w_rist_connect, w_rist_connection_status_callback, w_rist_ctx, w_rist_data_block, w_rist_fini,
    w_rist_free_data_block, w_rist_get_data_block, w_rist_get_data_block_len, w_rist_init,
    w_rist_init_data_block, w_rist_log_callback, w_rist_out_of_band_callback, w_rist_read,
    w_rist_set_auth_handler, w_rist_set_conn_status_callback, w_rist_set_data_block,
    w_rist_set_out_of_band_callback, w_rist_set_receiver_data_callback, w_rist_set_stats_callback,
    w_rist_stats_callback, w_rist_write,
};
use crate::system::ffi::version::size_t;
use anyhow::{bail, Result};
use std::os::raw::c_void;

#[derive(Debug, Clone, Copy)]
#[repr(C)]
pub enum rist_connection_status {
    RIST_CONNECTION_ESTABLISHED = 0,
    RIST_CONNECTION_TIMED_OUT,
    RIST_CLIENT_CONNECTED,
    RIST_CLIENT_TIMED_OUT,
}

#[derive(Debug, Clone, Copy)]
#[repr(C)]
pub enum rist_ctx_mode {
    RIST_SENDER_MODE = 0,
    RIST_RECEIVER_MODE = 1,
}

#[derive(Debug, Clone, Copy)]
#[repr(C)]
pub enum rist_profile {
    RIST_PROFILE_SIMPLE = 0,
    RIST_PROFILE_MAIN = 1,
    RIST_PROFILE_ADVANCED = 2,
}

#[derive(Debug, Clone, Copy)]
#[repr(C)]
pub enum rist_log_level {
    RIST_LOG_DISABLE = -1,
    RIST_LOG_ERROR = 3,
    RIST_LOG_WARN = 4,
    RIST_LOG_NOTICE = 5,
    RIST_LOG_INFO = 6,
    RIST_LOG_DEBUG = 7,
    RIST_LOG_SIMULATE = 100,
}
impl rist_log_level {
    #[must_use]
    pub const fn from_i32(value: i32) -> Self {
        match value {
            3 => Self::RIST_LOG_ERROR,
            4 => Self::RIST_LOG_WARN,
            5 => Self::RIST_LOG_NOTICE,
            6 => Self::RIST_LOG_INFO,
            7 => Self::RIST_LOG_DEBUG,
            100 => Self::RIST_LOG_SIMULATE,
            _ => Self::RIST_LOG_DISABLE,
        }
    }
}

#[derive(Clone)]
pub struct rist_data_block {
    pub block: w_rist_data_block,
}

impl Default for rist_data_block {
    fn default() -> Self {
        Self {
            block: std::ptr::null_mut(),
        }
    }
}

impl Drop for rist_data_block {
    fn drop(&mut self) {
        if !self.block.is_null() {
            unsafe { w_rist_free_data_block(&mut self.block) };
        }
    }
}

impl rist_data_block {
    /// # Errors
    ///
    /// TODO: add error description
    pub fn new() -> Result<Self> {
        let mut rist_block = Self {
            block: std::ptr::null_mut(),
        };
        let ret = unsafe { w_rist_init_data_block(&mut rist_block.block) };
        if ret == 0 {
            return Ok(rist_block);
        }
        bail!("could not allocate memory for rist_data_block")
    }

    /// # Errors
    ///
    /// TODO: add error description
    pub fn from(p_data_block: w_rist_data_block) -> Result<Self> {
        let mut rist_block = Self {
            block: p_data_block,
        };
        let ret = unsafe { w_rist_init_data_block(&mut rist_block.block) };
        if ret == 0 {
            return Ok(rist_block);
        }
        bail!("could not allocate memory for rist_data_block")
    }

    /// # Errors
    ///
    /// TODO: add error description
    pub fn read(&mut self) -> Result<&[u8]> {
        const TRACE: &str = "rist::read";
        unsafe {
            let ptr = w_rist_get_data_block(self.block);
            let len = w_rist_get_data_block_len(self.block);
            let len_usize = usize::try_from(len).unwrap_or(0);
            if len_usize > 0 {
                return Ok(std::slice::from_raw_parts(ptr as *mut u8, len_usize));
            }
            bail!("invalid len size. trace info {}", TRACE)
        }
    }

    pub fn set(&mut self, p_data: &[u8]) {
        unsafe {
            w_rist_set_data_block(
                self.block,
                p_data.as_ptr().cast::<std::ffi::c_void>(),
                p_data.len() as size_t,
            );
        };
    }
}

#[derive(Clone)]
pub struct rist {
    pub ctx: w_rist_ctx,
    pub url: String,
    pub mode: rist_ctx_mode,
    pub profile: rist_profile,
    pub loss_percentage: u16,
    pub log_level: rist_log_level,
}

impl Drop for rist {
    fn drop(&mut self) {
        //drop rist context
        unsafe { w_rist_fini(&mut self.ctx) };
    }
}

impl rist {
    /// # Errors
    ///
    /// TODO: add error description
    pub fn new(
        p_mode: rist_ctx_mode,
        p_profile: rist_profile,
        p_loss_percentage: u16,
        p_log_level: rist_log_level,
        p_log_callback: w_rist_log_callback,
    ) -> Result<Self> {
        let mut obj = Self {
            ctx: std::ptr::null_mut(),
            url: "".to_owned(),
            mode: p_mode,
            profile: p_profile,
            loss_percentage: p_loss_percentage,
            log_level: p_log_level,
        };

        // create sender/receiver context
        let ret = unsafe {
            w_rist_init(
                &mut obj.ctx,
                obj.mode as i32,
                obj.profile as i32,
                obj.loss_percentage,
                obj.log_level as i32,
                p_log_callback,
            )
        };
        match ret {
            0 => Ok(obj),
            _ => {
                bail!("could not create rist {:?}", obj.mode)
            }
        }
    }

    /// # Errors
    ///
    /// TODO: add error description
    pub fn connect(&mut self, p_url: &str) -> Result<()> {
        self.url = p_url.to_string();
        let ret = unsafe { w_rist_connect(self.ctx, self.url.as_ptr().cast::<i8>()) };
        match ret {
            0 => Ok(()),
            _ => {
                bail!("could not connect to {}", self.url)
            }
        }
    }

    /// # Errors
    ///
    /// TODO: add error description
    /// # Safety
    ///
    /// This function use void* for arg
    pub unsafe fn set_auth_handler(
        &self,
        p_arg: *mut c_void,
        p_on_auth_connect_cb: w_rist_auth_connect_callback,
        p_on_auth_disconnect_cb: w_rist_auth_disconnect_callback,
    ) -> Result<()> {
        let ret = w_rist_set_auth_handler(
            self.ctx,
            p_arg,
            p_on_auth_connect_cb,
            p_on_auth_disconnect_cb,
        );
        match ret {
            0 => Ok(()),
            _ => {
                bail!("could not set auth handler for {}", self.url)
            }
        }
    }

    /// # Errors
    ///
    /// TODO: add error description
    /// # Safety
    ///
    /// This function use void* for arg
    pub unsafe fn set_conn_status_callback(
        &self,
        p_arg: *mut c_void,
        p_on_callback: w_rist_connection_status_callback,
    ) -> Result<()> {
        let ret = w_rist_set_conn_status_callback(self.ctx, p_arg, p_on_callback);
        match ret {
            0 => Ok(()),
            _ => {
                bail!("could not set connection status callback for {}", self.url)
            }
        }
    }

    /// # Errors
    ///
    /// TODO: add error description
    /// # Safety
    ///
    /// This function use void* for arg
    pub unsafe fn set_out_of_band_callback(
        &self,
        p_arg: *mut c_void,
        p_on_callback: w_rist_out_of_band_callback,
    ) -> Result<()> {
        let ret = w_rist_set_out_of_band_callback(self.ctx, p_arg, p_on_callback);
        match ret {
            0 => Ok(()),
            _ => {
                bail!("could not set out of band callback for {}", self.url)
            }
        }
    }
    /// # Errors
    ///
    /// TODO: add error description
    /// # Safety
    ///
    /// This function use void* for arg
    pub unsafe fn set_stats_callback(
        &self,
        p_interval: i32,
        p_arg: *mut c_void,
        p_on_callback: w_rist_stats_callback,
    ) -> Result<()> {
        let ret = w_rist_set_stats_callback(self.ctx, p_interval, p_arg, p_on_callback);
        match ret {
            0 => Ok(()),
            _ => {
                bail!("could not set stats callback for {}", self.url)
            }
        }
    }

    /// # Errors
    ///
    /// TODO: add error description
    /// # Safety
    ///
    /// This function use void* for arg
    pub unsafe fn set_receiver_data_callback(
        &self,
        p_arg: *mut c_void,
        p_on_callback: w_receiver_data_callback,
    ) -> Result<()> {
        let ret = w_rist_set_receiver_data_callback(self.ctx, p_on_callback, p_arg);
        match ret {
            0 => Ok(()),
            _ => {
                bail!("could not set receiver callback for {}", self.url)
            }
        }
    }

    #[must_use]
    pub fn send(&self, p_data_block: &rist_data_block) -> i32 {
        unsafe { w_rist_write(self.ctx, p_data_block.block) }
    }

    pub fn read(&self, p_data_block: &mut rist_data_block, p_timeout: i32) -> i32 {
        unsafe { w_rist_read(self.ctx, &mut p_data_block.block, p_timeout) }
    }
}

File: wolf/sys/stream/rist.cpp
/*
        Project: Wolf Engine. Copyright 2014-2022 Pooya Eimandar
        https://github.com/WolfEngine/WolfEngine
*/
#include "rist.h"
#include <gsl/gsl> //NO-LINT
#include <librist/librist.h>
#include <rist-private.h>

struct w_rist_ctx_t
{
    rist_ctx* ctx;
    rist_peer* peer;
    rist_profile profile;
    uint16_t loss_percentage;
    rist_logging_settings* log;
};

int w_rist_init(w_rist_ctx* p_rist, enum rist_ctx_mode p_mode,
    enum rist_profile p_profile, uint16_t p_loss_percentage,
    enum rist_log_level p_log_level,
    w_rist_log_callback p_log_callback)
{
    constexpr auto size = sizeof(w_rist_ctx_t);
    auto* rist = gsl::owner<w_rist_ctx_t*>(calloc(1, size));
    if (rist == nullptr) 
    {
        return W_FAILURE;
    }

    rist->profile = p_profile;
    rist->loss_percentage = p_loss_percentage;

    // create rist context
    rist_ctx* ctx = nullptr;

    switch (p_mode) 
    {
    case rist_ctx_mode::RIST_SENDER_MODE: 
    {
        // create a log
        if (rist_logging_set(&rist->log, p_log_level, p_log_callback,
            (void*)("sender"), nullptr, stderr) != 0) 
        {
            return W_FAILURE;
        }
        // create a sender
        auto ret = rist_sender_create(&ctx, rist->profile, 0, rist->log);
        if (ret != 0) 
        {
            rist_log(rist->log, RIST_LOG_ERROR,
                "could not create rist sender context");
            return W_FAILURE;
        }
        if (p_loss_percentage > 0) 
        {
            ctx->sender_ctx->simulate_loss = true;
            ctx->sender_ctx->loss_percentage = p_loss_percentage * 10;
        }
        break;
    }
    case rist_ctx_mode::RIST_RECEIVER_MODE: 
    {
        // create a log
        if (rist_logging_set(&rist->log, p_log_level, p_log_callback,
            (void*)("receiver"), nullptr, stderr) != 0)
        {
            return W_FAILURE;
        }
        // create a receiver
        auto ret = rist_receiver_create(&ctx, rist->profile, rist->log);
        if (ret != 0)
        {
            rist_log(rist->log, rist_log_level::RIST_LOG_ERROR,
                "could not create rist receiver context");
            return W_FAILURE;
        }
        if (p_loss_percentage > 0)
        {
            ctx->receiver_ctx->simulate_loss = true;
            ctx->receiver_ctx->loss_percentage = p_loss_percentage * 10;
        }

        break;
    }
    };

    rist->ctx = ctx;
    *p_rist = rist;

    return W_SUCCESS;
}

int w_rist_connect(w_rist_ctx p_rist, const char* p_url)
{
    // rely on the library to parse the url
    rist_peer_config* peer_config = nullptr;
    if (rist_parse_address2(p_url, &peer_config)) {
        rist_log(p_rist->log, rist_log_level::RIST_LOG_ERROR,
            "could not parse peer options for receiver");
        return W_FAILURE;
    }

    if (rist_peer_create(p_rist->ctx, &p_rist->peer, peer_config) == -1) {
        rist_log(p_rist->log, rist_log_level::RIST_LOG_ERROR,
            "could not add peer connector to receiver");
        return W_FAILURE;
    }
    free((void*)peer_config);

    if (rist_start(p_rist->ctx) == -1) {
        rist_log(p_rist->log, rist_log_level::RIST_LOG_ERROR,
            "could not start rist receiver");
        return W_FAILURE;
    }

    return W_SUCCESS;
}

int w_rist_set_auth_handler(
    w_rist_ctx p_rist, void* p_arg,
    w_rist_auth_connect_callback p_on_auth_connect_cb,
    w_rist_auth_disconnect_callback p_on_auth_disconnect_cb)
{
    if (rist_auth_handler_set(p_rist->ctx, p_on_auth_connect_cb,
        p_on_auth_disconnect_cb, p_arg) != 0) 
    {
        rist_log(p_rist->log, rist_log_level::RIST_LOG_ERROR,
            "could not init rist auth handler\n");
        return W_FAILURE;
    }
    return W_SUCCESS;
}

int w_rist_set_conn_status_callback(
    w_rist_ctx p_rist, void* p_arg,
    w_rist_connection_status_callback p_on_connect_status_cb)
{
    // register callback for connection status
    if (rist_connection_status_callback_set(p_rist->ctx, p_on_connect_status_cb, p_arg) == -1)
    {
        rist_log(p_rist->log, RIST_LOG_ERROR,
            "could not initialize rist connection status callback\n");
        return W_FAILURE;
    }
    return W_SUCCESS;
}

int w_rist_set_out_of_band_callback(
    w_rist_ctx p_rist, void* p_arg,
    w_rist_out_of_band_callback p_on_out_of_band_cb)
{
    if (p_rist->profile == RIST_PROFILE_SIMPLE)
    {
        return W_FAILURE;
    }

    // register callback for connection status
    if (rist_oob_callback_set(p_rist->ctx, p_on_out_of_band_cb, p_arg) == -1)
    {
        rist_log(p_rist->log, RIST_LOG_ERROR,
            "could not add enable cout-of-band data\n");
        return W_FAILURE;
    }
    return W_SUCCESS;
}

int w_rist_set_stats_callback(w_rist_ctx p_rist, int p_interval, void* p_arg,
    w_rist_stats_callback p_on_stats_cb)
{
    // register callback for connection status
    if (rist_stats_callback_set(p_rist->ctx, p_interval, p_on_stats_cb, p_arg) == -1)
    {
        rist_log(p_rist->log, RIST_LOG_ERROR,
            "could not add enable stats callback\n");
        return W_FAILURE;
    }
    return W_SUCCESS;
}

int w_rist_write(w_rist_ctx p_rist, w_rist_data_block p_block)
{
    return rist_sender_data_write(p_rist->ctx, p_block);
}

int w_rist_read(w_rist_ctx p_rist, w_rist_data_block* p_block, int p_timeout) 
{
    return rist_receiver_data_read2(p_rist->ctx, p_block, p_timeout);
}

int w_rist_set_receiver_data_callback(w_rist_ctx p_rist,
    w_receiver_data_callback p_on_receive_cb,
    void* p_arg)
{
    if (rist_receiver_data_callback_set2(p_rist->ctx, p_on_receive_cb, &p_arg) != 0)
    {
        rist_log(p_rist->log, RIST_LOG_ERROR,
            "could not set data_callback pointer\n");
        return W_FAILURE;
    }
    return W_SUCCESS;
}

void w_rist_fini(w_rist_ctx* p_rist)
{
    if (p_rist == nullptr || (*p_rist) == nullptr)
    {
        return;
    }

    if ((*p_rist)->ctx)
    {
        rist_destroy((*p_rist)->ctx);
    }

    // use free instead of mi_free
    // there is a bug with mi_free on windows
    if ((*p_rist)->log)
    {
        free((*p_rist)->log);
    }

    free(*p_rist);
}